code/datums/elements/connect_loc.dm

```diff
@@ -0,0 +1,88 @@
+/// This element hooks a signal onto the loc the current object is on.
+/// When the object moves, it will unhook the signal and rehook it to the new object.
+/datum/element/connect_loc
+	element_flags = ELEMENT_BESPOKE | ELEMENT_DETACH
+	id_arg_index = 2
+
+	/// An assoc list of signal -> procpath to register to the loc this object is on.
+	var/list/connections
+
+	/// A list of everything with this element so that it can be tracked during ChangeTurf.
+	var/list/targets
+
+/datum/element/connect_loc/Attach(datum/target, list/connections)
+	. = ..()
+	if (!ismovable(target))
+		return ELEMENT_INCOMPATIBLE
+
+	src.connections = connections
+
+	LAZYADD(targets, target)
```

Comment:
Make this a keyed list of `turf = list(target)`

You don't need to worry about having a reference to a turf being changed, references to the old turf are valid references to the new turf replacing it.